proxy/src/connection.rs

```diff
@@ -254,6 +249,39 @@ impl Future for ConditionallyUpgradeServerToTls {
     }
 }
 
+impl ConditionallyUpgradeServerToTlsInner {
+    /// Polls the underliyng socket for more data and buffers it.
+    ///
+    /// The buffer is matched for a TLS client hello message.
+    ///
+    /// An error is returned if the underlying socket has closed.
+    fn poll_match_client_hello(&mut self) -> Poll<tls::conditional_accept::Match, io::Error> {
```

Comment:
I think this is the right place to put the code for handling 0-sized reads (rather than, say, adding a `Result` to the return type of `conditional_accept::match_client_hello` or something).

However, I do kind of wonder whether there are more places where we might read from a closed socket, get 0 bytes, and continue trying to read rather than noticing that the socket has closed, so perhaps we ought to be wrapping more sockets with this behaviour?

Reply:
Only in this file and tcp.rs. I think this is okay.

Reply:
Okay, SGTM.